packages/build/src/plugins_core/blobs_upload/utils.ts

```diff
@@ -0,0 +1,51 @@
+import fs from 'node:fs'
+import path from 'node:path'
+
+import { fdir } from 'fdir'
+
+const metadataPrefix = '$'
+const metadataSuffix = '.json'
+
+export const getBlobsDir = function ({ buildDir, publishDir }) {
+  const blobsPath = '.netlify/blobs/deploy'
+  const absoluteBlobsPath = path.resolve(buildDir, publishDir, blobsPath)
+  const blobsDir = path.normalize(absoluteBlobsPath)
+  return blobsDir
+}
+
+export const anyBlobsToUpload = function ({ buildDir, publishDir }) {
+  const blobsDir = getBlobsDir({ buildDir, publishDir })
+  const { files } = new fdir().onlyCounts().crawl(blobsDir).sync()
+  return files > 0
+}
+
+// Given output directory, find all file paths to upload excluding metadata files
+export async function getKeysToUpload(blobsDir): Promise<string[]> {
+  const files = await new fdir()
+    .withRelativePaths() // we want the relative path from the blobsDir
+    .filter((fpath) => !path.basename(fpath).startsWith(metadataPrefix))
+    .crawl(blobsDir)
+    .withPromise()
+
+  // normalize the path separators to all use the forward slash
+  return files.map((f) => f.split(path.sep).join('/'))
+}
+
+// Read a file and its metadata from the blobs directory
+export async function getFileWithMetadata(
+  blobsDir: string,
+  key: string,
+): Promise<{ data: Buffer; metadata: Record<string, string> }> {
+  const contentPath = path.join(blobsDir, key)
+  const dirname = path.dirname(key)
+  const basename = path.basename(key)
+  const metadataPath = path.join(blobsDir, dirname, `${metadataPrefix}${basename}${metadataSuffix}`)
+
+  const data = fs.readFileSync(contentPath)
+  let metadata = {}
+  if (fs.existsSync(metadataPath)) {
+    const metadataFile = fs.readFileSync(metadataPath, { encoding: 'utf8' })
```

Comment:
Good call! Will update.

Reply:
I've updated it in 3d21cdfa6 so it does async a bit better, however still using `existsSync` (hopefully it's a very fast check) since it's a bit nicer.
It'll probably benefit from using `fs.stat` (the promise version) and doing some extra checks though, right?